sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/HllCountMergePartialFn.java

```diff
@@ -0,0 +1,151 @@
+   * operation, and we are not able to create such "identity element" of type {@link
+   * HyperLogLogPlusPlus} without knowing the sketch precision and value type.
+   */
+  static final class HyperLogLogPlusPlusWrapper<T> {
+    @Nullable private HyperLogLogPlusPlus<T> hll;
+
+    private HyperLogLogPlusPlusWrapper(@Nullable HyperLogLogPlusPlus<T> hll) {
+      this.hll = hll;
+    }
+  }
+
+  /**
+   * Coder for the {@link HyperLogLogPlusPlusWrapper} class with generic type parameter {@code T}.
+   */
+  private static final class HyperLogLogPlusPlusWrapperCoder<T>
+      extends AtomicCoder<HyperLogLogPlusPlusWrapper<T>> {
+
+    private static final HyperLogLogPlusPlusWrapperCoder<?> INSTANCE =
+        new HyperLogLogPlusPlusWrapperCoder<>();
+
+    private final Coder<HyperLogLogPlusPlus<T>> nullableHllCoder =
+        NullableCoder.of(HyperLogLogPlusPlusCoder.of());
+
+    // Generic singleton factory pattern; the coder works for all HyperLogLogPlusPlusWrapper objects
+    // at runtime regardless of type T
+    @SuppressWarnings("unchecked")
+    static <T> HyperLogLogPlusPlusWrapperCoder<T> of() {
+      return (HyperLogLogPlusPlusWrapperCoder<T>) INSTANCE;
+    }
+
+    @Override
+    public void encode(HyperLogLogPlusPlusWrapper<T> value, OutputStream outStream)
+        throws IOException {
+      nullableHllCoder.encode(value.hll, outStream);
+    }
+
+    @Override
+    public HyperLogLogPlusPlusWrapper<T> decode(InputStream inStream) throws IOException {
+      return new HyperLogLogPlusPlusWrapper<T>(nullableHllCoder.decode(inStream));
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      nullableHllCoder.verifyDeterministic();
+    }
+  }
+
+  @Override
+  public Coder<HyperLogLogPlusPlusWrapper<T>> getAccumulatorCoder(
+      CoderRegistry registry, Coder<byte[]> inputCoder) {
+    return HyperLogLogPlusPlusWrapperCoder.of();
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> createAccumulator() {
+    return new HyperLogLogPlusPlusWrapper<>(null);
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> addInput(
+      HyperLogLogPlusPlusWrapper<T> accumulator, byte[] input) {
+    if (accumulator.hll == null) {
+      @SuppressWarnings("unchecked")
+      HyperLogLogPlusPlus<T> hll = (HyperLogLogPlusPlus<T>) HyperLogLogPlusPlus.forProto(input);
+      accumulator.hll = hll;
+    } else {
+      accumulator.hll.merge(input);
+    }
+    return accumulator;
+  }
+
+  @Override
+  public HyperLogLogPlusPlusWrapper<T> mergeAccumulators(
+      Iterable<HyperLogLogPlusPlusWrapper<T>> accumulators) {
+    HyperLogLogPlusPlusWrapper<T> merged = createAccumulator();
+    for (HyperLogLogPlusPlusWrapper<T> accumulator : accumulators) {
+      if (accumulator.hll == null) {
+        continue;
+      }
+      if (merged.hll == null) {
+        // Cannot set merged.hll to accumulator.hll directly because we shouldn't mutate accumulator
```

Comment:
Promoting to its own comment thread, as otherwise it gets lost in the already resolved comment thread: 
Would it make sense to move this comment to right on top of line 143? I was asking myself the exactly same thing as @amaliujia, and also overlooked this comment further up.
P.S. the serialize-then-deserialize step is fairly expensive, if we ever find a workaround for it in the future.

Reply:
> CombineFn#mergeAccumulators javadoc says:
"May modify any of the argument accumulators. May return a fresh accumulator, or may return one of the (modified) argument accumulators."
That seems to contradict the "we shouldn't mutate accumulator"?

Yes they are contradictory, and the doc you read is outdated. I did not find the javadoc you mentioned. Could you please share a link to it? Thank you!

> If we indeed cannot mutate them, I would move this comment to right on top of line 143

Done! (I think you mean line 134)

> P.S. the serialize-then-deserialize step is fairly expensive, if we ever find a workaround for it in the future.
Yes I agree. But it is good that this serialize-then-deserialize step will not be called very often (number of calls = number of "identity elements").

Reply:
That is up-to-date. The first accumulator in the `Iterable` can be modified for efficiency. But in our case we did not modify any of them.